File: dfa.py
import sys
import re
from typing import Dict, Set, Tuple

COMMENT_RE = re.compile(r"^\s*#")

def parse_conf(path: str):
    states: Set[str] = set()
    alphabet: Set[str] = set()
    start: str | None = None
    accept: Set[str] = set()
    delta: Dict[Tuple[str, str], str] = {}

    with open(path, 'r', encoding='utf-8') as f:
        lines = [ln.strip() for ln in f]

    # filtra comentarios y vacíos
    lines = [ln for ln in lines if ln and not COMMENT_RE.match(ln)]

    i = 0
    while i < len(lines):
        ln = lines[i]
        if ln.lower().startswith('states:'):
            vals = ln.split(':', 1)[1]
            states = {s.strip() for s in vals.split(',') if s.strip()}
        elif ln.lower().startswith('alphabet:'):
            vals = ln.split(':', 1)[1]
            alpha_list = [s.strip() for s in vals.split(',') if s.strip()]
            for s in alpha_list:
                if len(s) != 1:
                    raise ValueError(f"Símbolo '{s}' debe ser de un carácter")
            alphabet = set(alpha_list)
        elif ln.lower().startswith('start:'):
            start = ln.split(':', 1)[1].strip()
        elif ln.lower().startswith('accept:'):
            vals = ln.split(':', 1)[1]
            accept = {s.strip() for s in vals.split(',') if s.strip()}
        elif ln.lower().startswith('transitions:'):
            i += 1
            while i < len(lines):
                tln = lines[i].strip()
                if not tln or COMMENT_RE.match(tln):
                    i += 1
                    continue
                # formato: q, a -> p
                parts = tln.split('->')
                if len(parts) != 2:
                    raise ValueError(f"Transición inválida: {tln}")
                left, right = parts[0].strip(), parts[1].strip()
                if ',' not in left:
                    raise ValueError(f"Transición inválida (falta coma): {tln}")
                st, sym = [x.strip() for x in left.split(',', 1)]
                if len(sym) != 1:
                    raise ValueError(f"Símbolo '{sym}' debe ser de un carácter")
                delta[(st, sym)] = right
                i += 1
            break  # fin del archivo
        i += 1

    if not states:
        raise ValueError("'states' vacío o ausente")
    if not alphabet:
        raise ValueError("'alphabet' vacío o ausente")
    if start is None or start not in states:
        raise ValueError("'start' inválido o ausente")
    if not accept.issubset(states):
        raise ValueError("Algún estado en 'accept' no pertenece a 'states'")

    return states, alphabet, start, accept, delta

def simulate(word: str, start: str, accept: Set[str], delta: Dict[Tuple[str, str], str], alphabet: Set[str]):
    q = start
    for ch in word:
        if ch not in alphabet:
            return False
        nxt = delta.get((q, ch))
        if nxt is None:
            return False
        q = nxt
    return q in accept

def main():
    if len(sys.argv) not in (1, 3):
        print("Uso: python dfa.py [Conf.txt Cadenas.txt]")
        sys.exit(1)
    conf_path = sys.argv[1] if len(sys.argv) == 3 else 'Conf.txt'
    words_path = sys.argv[2] if len(sys.argv) == 3 else 'Cadenas.txt'

    states, alphabet, start, accept, delta = parse_conf(conf_path)

    with open(words_path, 'r', encoding='utf-8') as f:
        for raw in f:
            s = raw.strip('\n')
            if not s.strip() or COMMENT_RE.match(s):
                continue
            ok = simulate(s.strip(), start, accept, delta, alphabet)
            print(f"{s}: {'ACEPTADA' if ok else 'RECHAZADA'}")

if __name__ == '__main__':
    main()

File: dfa.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>

#define MAX_STATES 128
#define MAX_ALPHA 128
#define MAX_LINE 1024
#define MAX_NAME 128

static char states[MAX_STATES][MAX_NAME];
static int n_states = 0;
static char alphabet[MAX_ALPHA]; // símbolos de 1 char
static int n_alpha = 0;
static int start_idx = -1;
static int accept_mask[MAX_STATES];
static int trans[MAX_STATES][MAX_ALPHA]; // -1 si no definida

static int is_comment_or_empty(const char *s){
    while (*s && isspace((unsigned char)*s)) s++;
    return (*s == '\0' || *s == '#');
}

static int trim(char *s){
    int len = (int)strlen(s);
    while (len>0 && (s[len-1]=='\n' || s[len-1]=='\r')) s[--len] = '\0';
    int i=0; while (isspace((unsigned char)s[i])) i++;
    if(i>0) memmove(s, s+i, len - i + 1);
    len = (int)strlen(s);
    while (len>0 && isspace((unsigned char)s[len-1])) s[--len] = '\0';
    return len;
}

static int find_state(const char *name){
    for (int i=0;i<n_states;i++) if (strcmp(states[i], name)==0) return i;
    return -1;
}

static int add_state(const char *name){
    if (n_states >= MAX_STATES) { fprintf(stderr, "Demasiados estados\n"); exit(1);}    
    strncpy(states[n_states], name, MAX_NAME-1);
    states[n_states][MAX_NAME-1] = '\0';
    return n_states++;
}

static int find_symbol(char c){
    for (int i=0;i<n_alpha;i++) if (alphabet[i]==c) return i;
    return -1;
}

static void parse_csv_states(char *vals){
    char *tok = strtok(vals, ",");
    while (tok){
        trim(tok);
        if (*tok){
            if (find_state(tok) == -1) add_state(tok);
        }
        tok = strtok(NULL, ",");
    }
}

static void parse_csv_alphabet(char *vals){
    char *tok = strtok(vals, ",");
    while (tok){
        trim(tok);
        if (*tok){
            if (strlen(tok) != 1){ fprintf(stderr, "Símbolo '%s' debe ser de 1 carácter\n", tok); exit(1);}            
            if (n_alpha >= MAX_ALPHA){ fprintf(stderr, "Alfabeto demasiado grande\n"); exit(1);}            
            if (find_symbol(tok[0]) == -1) alphabet[n_alpha++] = tok[0];
        }
        tok = strtok(NULL, ",");
    }
}

static void init_trans(){
    for (int i=0;i<MAX_STATES;i++) for (int j=0;j<MAX_ALPHA;j++) trans[i][j] = -1;
    for (int i=0;i<MAX_STATES;i++) accept_mask[i] = 0;
}

static void parse_conf(const char *path){
    FILE *f = fopen(path, "r");
    if (!f){ perror("Conf.txt"); exit(1);}    
    init_trans();

    char line[MAX_LINE];
    int in_transitions = 0;

    while (fgets(line, sizeof(line), f)){
        trim(line);
        if (is_comment_or_empty(line)) continue;

        if (!in_transitions){
            if (strncasecmp(line, "states:", 7) == 0){
                char *vals = line + 7; while (*vals==':'||isspace((unsigned char)*vals)) vals++;
                parse_csv_states(vals);
            } else if (strncasecmp(line, "alphabet:", 9) == 0){
                char *vals = line + 9; while (*vals==':'||isspace((unsigned char)*vals)) vals++;
                parse_csv_alphabet(vals);
            } else if (strncasecmp(line, "start:", 6) == 0){
                char *val = line + 6; while (*val==':'||isspace((unsigned char)*val)) val++;
                trim(val);
                int idx = find_state(val);
                if (idx == -1){ fprintf(stderr, "Estado start '%s' no está en states\n", val); exit(1);}                
                start_idx = idx;
            } else if (strncasecmp(line, "accept:", 7) == 0){
                char buf[MAX_LINE]; strncpy(buf, line + 7, sizeof(buf)-1); buf[sizeof(buf)-1]='\0';
                char *tok = strtok(buf, ",");
                while (tok){
                    trim(tok);
                    if (*tok){
                        int idx = find_state(tok);
                        if (idx == -1){ fprintf(stderr, "Estado de accept '%s' no está en states\n", tok); exit(1);}                        
                        accept_mask[idx] = 1;
                    }
                    tok = strtok(NULL, ",");
                }
            } else if (strncasecmp(line, "transitions:", 11) == 0){
                in_transitions = 1;
            } else {
                fprintf(stderr, "Línea desconocida: %s\n", line);
                exit(1);
            }
        } else {
            // formato: estado,simbolo->destino
            char *arrow = strstr(line, "->");
            if (!arrow){ fprintf(stderr, "Transición inválida: %s\n", line); exit(1);}            
            *arrow = '\0';
            char *right = arrow + 2;
            trim(right);

            char left[MAX_LINE]; strncpy(left, line, sizeof(left)-1); left[sizeof(left)-1]='\0';
            char *comma = strchr(left, ',');
            if (!comma){ fprintf(stderr, "Transición inválida (falta coma): %s\n", line); exit(1);}            
            *comma = '\0';
            char *st = left; trim(st);
            char *symstr = comma + 1; trim(symstr);
            if (strlen(symstr) != 1){ fprintf(stderr, "Símbolo '%s' debe ser de 1 carácter\n", symstr); exit(1);}            
            char sym = symstr[0];

            int qi = find_state(st);
            int si = find_symbol(sym);
            int qj = find_state(right);
            if (qi==-1 || si==-1 || qj==-1){
                fprintf(stderr, "Transición con estado/símbolo desconocido: %s,%c->%s\n", st, sym, right);
                exit(1);
            }
            trans[qi][si] = qj;
        }
    }

    fclose(f);
    if (n_states==0 || n_alpha==0 || start_idx<0){
        fprintf(stderr, "Configuración incompleta (states/alphabet/start)\n");
        exit(1);
    }
}

static int simulate_line(const char *s){
    int q = start_idx;
    for (const char *p=s; *p; ++p){
        char c = *p;
        if (c=='\n' || c=='\r') break;
        if (isspace((unsigned char)c)) continue; // opcional: ignorar espacios
        int si = find_symbol(c);
        if (si==-1) return 0; // símbolo fuera del alfabeto
        int qn = trans[q][si];
        if (qn < 0) return 0; // transición no definida
        q = qn;
    }
    return accept_mask[q];
}

int main(int argc, char **argv){
    const char *conf = (argc==3) ? argv[1] : "Conf.txt";
    const char *words = (argc==3) ? argv[2] : "Cadenas.txt";
    parse_conf(conf);

    FILE *f = fopen(words, "r");
    if (!f){ perror("Cadenas.txt"); return 1; }

    char line[MAX_LINE];
    while (fgets(line, sizeof(line), f)){
        char tmp[MAX_LINE];
        strncpy(tmp, line, sizeof(tmp)-1); tmp[sizeof(tmp)-1]='\0';
        trim(tmp);
        if (is_comment_or_empty(tmp)) continue;
        int ok = simulate_line(tmp);
        printf("%s: %s\n", tmp, ok?"ACEPTADA":"RECHAZADA");
    }

    fclose(f);
    return 0;
}
